src/main/java/sirius/web/security/Permissions.java

```diff
@@ -49,82 +49,100 @@
 
         for (Extension ext : Sirius.getSettings().getExtensions("security.profiles")) {
-            Set<String> profile = compileProfile(ext);
-            profiles.put(ext.getId(), profile);
+            profiles.add(compileProfile(ext));
         }
 
         profilesCache = profiles;
     }
 
-    private static Set<String> compileProfile(Extension ext) {
-        Set<String> permissions = Sets.newTreeSet();
+    private static Profile compileProfile(Extension ext) {
+        Profile profile = new Profile(ext.getId());
+
         for (Map.Entry<String, Object> permission : ext.getContext().entrySet()) {
-            if (Value.of(permission.getValue()).asBoolean()) {
-                permissions.add(permission.getKey());
+            if (Boolean.TRUE.equals(permission.getValue())) {
+                profile.addPermission(permission.getKey());
             }
-        }
-
-        return permissions;
-    }
-
-    private static void expand(String role, Set<String> result) {
-        if (!result.contains(role)) {
-            result.add(role);
-            for (String subRole : getProfile(role)) {
-                expand(subRole, result);
+            if (Boolean.FALSE.equals(permission.getValue())) {
+                profile.removePermission(permission.getKey());
             }
         }
+
+        return profile;
     }
 
     /**
-     * Expands all permission profiles to obtain the effective set of permissions for a given list or permission and
+     * Expands all permission profiles to obtain the effective set of permissions for a set list or permission and
      * profile names.
      *
-     * @param roles the list of permissions and or profiles to expand
-     * @return an effective list of permissions based on the profiles defined in <tt>security.profiles</tt>
+     * @param permissions the set of permissions and or profiles to expand
      */
-    public static Set<String> applyProfiles(Collection<String> roles) {
-        Set<String> result = Sets.newTreeSet();
-        for (String role : roles) {
-            expand(role, result);
+    public static void applyProfiles(Set<String> permissions) {
+        for (Profile profile : getProfiles()) {
+            profile.apply(permissions);
         }
-        return result;
     }
 
     /**
-     * Expands all permission profiles just like {@link #applyProfiles(java.util.Collection)}. Also all public roles
+     * Expands all permission profiles just like {@link #applyProfiles(java.util.Set)}. Also all public roles
      * defined in <tt>security.publicRoles</tt> are included to the roles set before profiles are expanded.
      *
      * @param roles the list of permissions and or profiles to expand
      * @return an effective list of permissions based on the profiles defined in <tt>security.profiles</tt> and the
      * public roles defined in <tt>security.publicRoles</tt>
      */
     public static Set<String> applyProfilesAndPublicRoles(Collection<String> roles) {
```

Comment:
this method signature should probably be changed too, to work like applyProfiles, given the similiar naming